docs/runtime-contract.md

```diff
@@ -91,9 +91,11 @@ In a highly-shared environment, containers might experience the following:
 
 ### Lifecycle
 
-- The container MAY be killed when the container is inactive. Serverless
-  computation relies on inbound requests to determine container activeness. The
-  container is sent a `SIGTERM` signal when it is killed via the
+- The container MAY be killed when the container is inactive. 
+  The mechanism by which container activeness is determined is an 
+  implementation detail. One common approach is to assume presence of 
```

Comment:
It looks like this is only changing non-requirements phrases (i.e. "container MAY be killed" is not changed, only the guidance around determining inactivity).

On a first reading, I don't see any behaviors that would be prohibited in one phrasing but allowed in the other -- in particular, if we say "The container MAY be killed when the container is inactive", it's perfectly legal to **not** kill the container when it is inactive, or to take other information into account before deciding whether or not to kill the container.

Given that this is a MAY, would converting it into an inverse like the following cause you any problems:

> The container SHOULD NOT be killed while at least one HTTP request is in flight.

Reply:
I believe the key wording is `relies on inbound requests to determine container activeness` - that's the bit that needs to be relaxed.

Reply:
I think we've agreed elsewhere that requests in flight **are** a signal of container activeness. However, there are also potentially **other** signals of container activeness.

How about:

> The container MAY be killed when the container is inactive. Containers MUST be considered "active" while they are handling at least one request, but other conditions MAY also be used to determine that a container is active. The container is sent a `SIGTERM`...


I'm fine with adding additional "is active" conditions which prevent termination; I'm more concerned with adding termination cases which make it difficult for developers to be certain that their request will be processed under normal operation.

Reply:
Can live with it, but I do have a slight concern with the phrase "at least one request" because "request" could be interpreted as "http request".  Would "job" (or "task") be an ok substitute ?

Reply:
The intent is to say:

IF at least one HTTP request is in flight, the container is active.
IF zero HTTP requests are in flight, the container MAY be active or not active.

Since we don't have an agreed-on mechanism yet for describing "job" status in this doc, I'd feel strange using that language here.

I think it's fine if people want to experiment with async workloads by having the execution environment keep track of overhanging-tail work (work that runs after the HTTP response has been returned), and I don't want to block it.

I also want to be clear about how activity is measured, so without a clear spec around how long the overhanging tail is, the best option I can see is to say "a container may be active at times when it is handling zero http requests".

Reply:
As an example alternate mechanism for tracking activity, you could imagine a response header `Execution-Grace-Time` that indicates that the container should continue to be considered active until the specified time has been reached. (This would compose nicely across multiple requests; in the server side, this could be tracked as a single "active high water mark", or could be tracked using concurrency mechanisms to limit the number of grace period processing items in flight alongside the number of active HTTP requests.)